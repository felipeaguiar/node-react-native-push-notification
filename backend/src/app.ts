import express from 'express';
import cors from 'cors';
import morgan from 'morgan';
import mustacheExpress from 'mustache-express';
import path from 'path';
import routes from './routes';
import handler from './middleware/handler';
import logger from './config/logger';
import moment from 'moment';

const app = express();

app.use(cors({
  exposedHeaders: [
    'X-Count'
  ]
}));

morgan.token('date', () => moment().format());
app.use(morgan('combined', { stream: { write: message => logger.info(message) } }));

app.use(express.json());
app.use(express.urlencoded({ extended: true }));

app.use(routes);
app.use(handler);

app.engine('html', mustacheExpress());
app.set('view engine', 'html');
app.set('views', path.join(__dirname, '/view'));

export default app;
