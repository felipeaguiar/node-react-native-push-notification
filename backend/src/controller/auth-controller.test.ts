import jwtDecode from 'jwt-decode';
import db from '../test/db';
import request from 'supertest';
import app from '../app';

beforeAll(async () => {
  await db.create('auth');
});

afterAll(async () => {
  await db.close();
});

it('deve cadastrar usuário com sucesso', async () => {
  expect.assertions(3);

  const response = await request(app)
    .post('/auth/register')
    .send({
      nome: 'Felipe Aguiar',
      email: 'felipe@mail.com',
      senha: '123456'
    });

  expect(response.status).toBe(200);

  expect(response.body).toStrictEqual({
    token: expect.any(String),
    usuario: {
      id: expect.any(Number),
      version: expect.any(Number),
      nome: 'Felipe Aguiar',
      papel: 'usuario',
      email: 'felipe@mail.com'
    }
  });

  const dump = await db.dump();
  expect(dump).toMatchSnapshot();
});

it('deve validar se o usuário está tentando cadastrar um usuário já existente', async () => {
  expect.assertions(2);

  const response = await request(app)
    .post('/auth/register')
    .send({
      nome: 'Felipe Aguiar',
      email: 'felipe@mail.com',
      senha: '123456'
    });

  expect(response.status).toBe(400);

  expect(response.body).toStrictEqual({
    status: 400,
    error: expect.objectContaining({
      message: expect.any(String)
    })
  });
});

it('deve falhar ao tentar cadastrar um usuário com dados inválidos', async () => {
  expect.assertions(3);

  const response = await request(app)
    .post('/auth/register')
    .send({
      nome: 'Fel',
      email: 'mail.com',
      senha: '1234'
    });

  expect(response.status).toBe(400);

  expect(response.body).toStrictEqual({
    status: 400,
    error: expect.arrayContaining([
      expect.objectContaining({
        message: expect.any(String),
        context: expect.objectContaining({ key: 'nome' })
      }),
      expect.objectContaining({
        message: expect.any(String),
        context: expect.objectContaining({ key: 'email' })
      }),
      expect.objectContaining({
        message: expect.any(String),
        context: expect.objectContaining({ key: 'senha' })
      })
    ])
  });

  expect(response.body.error).toHaveLength(3);
});

it('deve autenticar um usuário e receber o token de acesso válido', async () => {
  expect.assertions(3);

  const response = await request(app)
    .post('/auth/login')
    .send({
      email: 'felipe@mail.com',
      senha: '123456'
    });

  expect(response.status).toBe(200);

  expect(response.body).toStrictEqual({
    usuario: {
      id: expect.any(Number),
      version: expect.any(Number),
      nome: 'Felipe Aguiar',
      email: 'felipe@mail.com',
      papel: 'usuario'
    },
    token: expect.any(String)
  });

  const token = jwtDecode(response.body.token);

  expect(token).toStrictEqual({
    id: expect.any(Number),
    nome: expect.any(String),
    papel: expect.any(String),
    iat: expect.any(Number),
    exp: expect.any(Number)
  });
});

it('deve falhar ao tentar autenticar com email e senha inválidos', async () => {
  expect.assertions(2);

  const response = await request(app)
    .post('/auth/login')
    .send({
      email: 'mail.com',
      senha: '1234'
    });

  expect(response.status).toBe(400);

  expect(response.body).toStrictEqual({
    status: 400,
    error: expect.arrayContaining([
      expect.objectContaining({
        message: expect.any(String),
        context: expect.objectContaining({ key: 'email' })
      }),
      expect.objectContaining({
        message: expect.any(String),
        context: expect.objectContaining({ key: 'senha' })
      })
    ])
  });
});

it('deve falhar ao tentar autenticar com uma senha incorreta', async () => {
  expect.assertions(2);

  const response = await request(app)
    .post('/auth/login')
    .send({
      email: 'felipe@mail.com',
      senha: '654321'
    });

  expect(response.status).toBe(400);

  expect(response.body).toStrictEqual({
    status: 400,
    error: expect.objectContaining({
      message: expect.any(String)
    })
  });
});
