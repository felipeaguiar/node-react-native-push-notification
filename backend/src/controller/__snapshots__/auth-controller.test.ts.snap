// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`deve cadastrar usu√°rio com sucesso 1`] = `
Object {
  "Usuario": Array [
    Usuario {
      "email": "felipe@mail.com",
      "id": 1,
      "nome": "Felipe Aguiar",
      "papel": "usuario",
      "version": 1,
    },
  ],
}
`;
